File: packages/template-android/app/src/main/java/com/rayact/engine/RayactPlatformRegistry.kt
package com.rayact.engine

import android.content.Context
import android.view.View
import java.util.concurrent.ConcurrentHashMap

fun interface RayactPlatformModule {
    fun invoke(method: String, payload: ByteArray, completion: (Result<ByteArray>) -> Unit)
}

fun interface RayactPlatformViewFactory {
    fun create(context: Context, nodeId: Int, props: Map<String, Any?>): View
}

interface RayactPlatformModuleRegistration {
    fun register(context: Context, registry: RayactPlatformRegistry)
}

/**
 * Application-owned registry populated by generated autolinking code.
 * The engine package contains no optional module or platform-view implementation.
 */
class RayactPlatformRegistry {
    private val modules = ConcurrentHashMap<String, RayactPlatformModule>()
    private val viewFactories = ConcurrentHashMap<String, RayactPlatformViewFactory>()

    fun registerModule(name: String, module: RayactPlatformModule) {
        require(name.isNotBlank()) { "Platform module name must not be blank" }
        check(modules.putIfAbsent(name, module) == null) { "Platform module '$name' already registered" }
    }

    fun registerViewFactory(kind: String, factory: RayactPlatformViewFactory) {
        require(kind.isNotBlank()) { "Platform view kind must not be blank" }
        check(viewFactories.putIfAbsent(kind, factory) == null) { "Platform view '$kind' already registered" }
    }

    fun invoke(name: String, method: String, payload: ByteArray, completion: (Result<ByteArray>) -> Unit) {
        val module = modules[name]
        if (module == null) {
            completion(Result.failure(IllegalStateException("Platform module '$name' is not installed")))
            return
        }
        module.invoke(method, payload, completion)
    }

    fun createView(kind: String, context: Context, nodeId: Int, props: Map<String, Any?>): View? =
        viewFactories[kind]?.create(context, nodeId, props)

    fun hasModule(name: String): Boolean = modules.containsKey(name)
    fun hasViewFactory(kind: String): Boolean = viewFactories.containsKey(kind)
}

File: packages/template-ios/RayactPlatformRegistry.swift
import UIKit

typealias RayactPlatformCompletion = (Result<Data, Error>) -> Void
typealias RayactPlatformModule = (_ method: String, _ payload: Data, _ completion: @escaping RayactPlatformCompletion) -> Void
typealias RayactPlatformViewFactory = (_ nodeId: Int, _ properties: [String: Any]) -> UIView

protocol RayactPlatformModuleRegistration {
    func register(with registry: RayactPlatformRegistry)
}

/// Application-owned registry populated by generated autolinking code.
final class RayactPlatformRegistry {
    private var modules: [String: RayactPlatformModule] = [:]
    private var viewFactories: [String: RayactPlatformViewFactory] = [:]
    private let lock = NSLock()

    func registerModule(_ name: String, invoke: @escaping RayactPlatformModule) {
        precondition(!name.isEmpty, "Platform module name must not be empty")
        lock.lock()
        defer { lock.unlock() }
        precondition(modules[name] == nil, "Platform module '\(name)' already registered")
        modules[name] = invoke
    }

    func registerViewFactory(_ kind: String, factory: @escaping RayactPlatformViewFactory) {
        precondition(!kind.isEmpty, "Platform view kind must not be empty")
        lock.lock()
        defer { lock.unlock() }
        precondition(viewFactories[kind] == nil, "Platform view '\(kind)' already registered")
        viewFactories[kind] = factory
    }

    func invoke(_ name: String, method: String, payload: Data, completion: @escaping RayactPlatformCompletion) {
        lock.lock()
        let module = modules[name]
        lock.unlock()
        guard let module else {
            completion(.failure(NSError(
                domain: "RayactPlatformModule",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Platform module '\(name)' is not installed"]
            )))
            return
        }
        module(method, payload, completion)
    }

    func makeView(_ kind: String, nodeId: Int, properties: [String: Any]) -> UIView? {
        lock.lock()
        let factory = viewFactories[kind]
        lock.unlock()
        return factory?(nodeId, properties)
    }

    func hasModule(_ name: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return modules[name] != nil
    }
}
